pkg/didcomm/transport/http/outbound.go

```diff
@@ -0,0 +1,170 @@
+
+import (
+	"bytes"
+	"crypto/tls"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"path/filepath"
+	"strings"
+	"time"
+
+	tlsCertPool "github.com/hyperledger/aries-framework-go/pkg/didcomm/transport/http/tls"
+	"github.com/pkg/errors"
+)
+
+const commContentType = "application/didcomm-envelope-enc"
+
+// OutboundCommHTTP is the HTTP transport implementation of CommTransport
+// it embeds an http.Server and has an http.Client instance
+type OutboundCommHTTP struct {
+	client *http.Client
+	cfg    *OutboundCommConfig
+}
+
+// OutboundCommConfig are the HTTP client config used by OutboundCommHTTP
+type OutboundCommConfig struct {
+	Timeout      time.Duration
+	CACertsPaths string
+	caCertPool   tlsCertPool.CertPool
+}
+
+// NewOutboundCommFromConfig creates a new instance of CommHTTP to Post requests to other Agents
+// To get a valid instance of OutboundCommHTTP, a cfg must be set
+// if config is nil, this function will throw an error
+func NewOutboundCommFromConfig(cfg *OutboundCommConfig) (*OutboundCommHTTP, error) {
+	var cl *http.Client
+	if cfg != nil {
+		var err error
+		cl, err = newHTTPClient(cfg)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		return nil, errors.New("config is empty, cannot create new HTTP transport")
+	}
+
+	cm := &OutboundCommHTTP{
+		client: cl,
+		cfg:    cfg,
+	}
+	return cm, nil
+}
+
+// NewOutboundCommFromClient creates a new instance of CommHTTP to Post requests to other Agents
+// To get a valid instance of OutboundCommHTTP, a client must be set
+// if client is nil, this function will throw an error
+func NewOutboundCommFromClient(client *http.Client) (*OutboundCommHTTP, error) {
+	if client == nil {
+		return nil, errors.New("client is empty, cannot create new HTTP transport")
+	}
+
+	cm := &OutboundCommHTTP{
+		client: client,
+	}
+	return cm, nil
+}
+
+// Send sends a2a exchange data via HTTP (client side)
+func (cs *OutboundCommHTTP) Send(data string, url string) (string, error) {
+	resp, err := cs.client.Post(url, commContentType, bytes.NewBuffer([]byte(data)))
+	if err != nil {
+		log.Printf("HTTP Transport - Error posting did envelope to agent at [%s]: %v", url, err)
+		return "", err
+	}
+
+	var respData string
+	if resp != nil {
+		isStatusSuccess := resp.StatusCode == http.StatusAccepted || resp.StatusCode == http.StatusOK
+		if !isStatusSuccess {
+			return "", errors.Errorf("Warning - Received non success POST HTTP status from agent at [%s]: status : %v", url, resp.Status)
+		}
+		// handle response
+		defer func() {
+			err := resp.Body.Close()
+			if err != nil {
+				log.Printf("HTTP Transport - Error closing response body: %v", err)
+			}
+		}()
+		buf := new(bytes.Buffer)
+		_, err := buf.ReadFrom(resp.Body)
+		if err != nil {
+			return "", err
+		}
+		respData = buf.String()
+	}
+	return respData, nil
+}
+
+// creates a new instance of HTTP transport as a client
+func newHTTPClient(cfg *OutboundCommConfig) (*http.Client, error) {
+	var err error
+	var caCertPool tlsCertPool.CertPool
```

Comment:
TLS Cert Pool builder logic does not belong in this package.

The client should accept an interface for a cert pool, and builder utilities should all be outside.

Reply:
will refactor to add options support and allow for a certpool option .. should this dynamic cert pool creation code move to a separate "utility" package? or removed from this PR then we can add it later, if needed?

Reply:
A dynamic cert pool package seems like it could be its own PR.